lambda/value_rec_compiler.ml

```diff
@@ -0,0 +1,778 @@
+    This file implements the translation from generic definitions to Lambda.
+
+    The first step occurs during typechecking, in [Value_rec_check]:
+    [Dynamic] bindings need to be compiled as normal let bindings. This file
+    mostly deals with the [Static] bindings.
+
+    The three phases in this module are the following:
+
+    - Sizing: we first classify the definitions by their size, which determines
+      the compilation strategy for each binding.
+
+    - Function lifting: we then apply a transformation from general function
+      definitions to syntactic functions accepted by [Lletrec].
+      Examples:
+      {[
+        let rec f x = f x (* Syntactic *)
+        let rec f = fun x -> f x (* Syntactic *)
+        let rec f = let g x = f x in g (* Not syntactic *)
+        let rec f = let a = ... in (fun x -> f x) (* Not syntactic *)
+      ]}
+
+    - Compilation: we finally combine all of this to produce a Lambda term
+      for the recursive bindings.
+*)
+
+open Lambda
+
+(** {1. Sizing} *)
+
+(* Simple blocks *)
+type block_size =
+  | Regular_block of int
+  | Float_record of int
+
+type size =
+  | Unreachable
+  (** Non-returning expressions, like [raise exn].
+      In [Value_rec_check], they would be classified as [Dynamic],
+      but some of those appear during translation to Lambda.
+      For example, in [let rec f = let [| x |] = ... in fun y -> x + y]
+      the inner let binding gets translated to code that raises
+      [Match_failure] for non-matching branches.
+      Tracking [Unreachable] explicitly allows us to recover the size
+      of the only non-raising branch. *)
+  | Constant
+  (** Constant values.
+      Can be either an integer-like constant ([0], ['a'], [None],
+      the empty list or the unit constructor), or a structured constant
+      (["hello"], [Some 1], ...).
+
+      Integer constants cannot be pre-allocated, so need their own
+      classification and compilation scheme (See {!Compilation} below).
+      Structured constants could fit into the [Block] category, but we
+      choose to reuse the [constant] classification to avoid sorting
+      through the [Lconst] definitions.
+      It also generates slightly better code. *)
+  | Function
+  (** Function definitions.
+      This includes more than just obvious, syntactic function definitions;
+      see {!Function Lifting} for details. *)
+  | Block of block_size
+  (** Allocated values of a fixed size.
+      This corresponds to expressions ending in a single obvious allocation,
+      but also some more complex expressions where the block is bound to
+      an intermediate variable before being returned.
+  *)
+
+type binding_size = (lambda_with_env, size) Lazy_backtrack.t
+and lambda_with_env = {
+  lambda : lambda;
+  env : binding_size Ident.Map.t;
+}
+
+let dynamic_size () =
+  Misc.fatal_error "letrec: No size found for Static binding"
+
+let join_sizes size1 size2 =
```

Comment:
From the name, I would expect `join_sizes` to not produce an error when `size1` is the same as `size2`

Reply:
I've added a comment explaining some of issues around joining sizes.